packages/beacon-node/src/chain/validation/attestation.ts

```diff
@@ -789,3 +788,45 @@ export function computeSubnetForSlot(shuffling: EpochShuffling, slot: number, co
   const committeesSinceEpochStart = shuffling.committeesPerSlot * slotsSinceEpochStart;
   return (committeesSinceEpochStart + committeeIndex) % ATTESTATION_SUBNET_COUNT;
 }
+
+/**
+ * Return fork-dependent seen attestation key
+ *   - for pre-electra, it's the AttestationData base64
+ *   - for electra and later, it's the AttestationData base64 + committeeBits base64
+ *
+ * we always have attDataBase64 from the IndexedGossipQueue, getAttDataFromAttestationSerialized() just for backward compatible when beaconAttestationBatchValidation is false
+ * TODO: remove beaconAttestationBatchValidation flag since the batch attestation is stable
+ */
+export function getSeenAttDataKeyFromGossipAttestation(
+  forkSeq: ForkSeq,
+  attestation: GossipAttestation
+): SeenAttDataKey | null {
+  const {attDataBase64, serializedData} = attestation;
+  if (forkSeq >= ForkSeq.electra) {
+    const attData = attDataBase64 ?? getAttDataFromAttestationSerialized(serializedData);
+    const committeeBits = getCommitteeBitsFromAttestationSerialized(serializedData);
+    return attData && committeeBits ? attDataBase64 + committeeBits : null;
+  }
+
+  // pre-electra
+  return attDataBase64 ?? getAttDataFromAttestationSerialized(serializedData);
+}
+
+/**
+ * Extract attestation data key from SignedAggregateAndProof Uint8Array to use cached data from SeenAttestationDatas
+ *   - for pre-electra, it's the AttestationData base64
+ *   - for electra and later, it's the AttestationData base64 + committeeBits base64
+ */
+export function getSeenAttDataKeyFromSignedAggregateAndProof(
```

Comment:
I think `packages/beacon-node/src/chain/validation/aggregateAndProof.ts` is a better place for this function

Reply:
I leave it there intentionally, AggregateAndProof validation shares a lot of utils in this file, and we want to maintain logic of these 2 functions together to make sure they are the same